### PR TITLE
Add basic support for using cvc5

### PR DESCRIPTION
**Pull request by parno**
At this point, we can run with cvc5 on all of our medium-scale Verus projects (although with more failures and generally slower performance). 

However,  the changes thus far do not:
1. Rename the various functions like `set_z3_param_u32` that embed Z3 in the name
2. Update `vargo` to know about cvc5
3. Update the test suite to use cvc5
4. Allow function-level rlimit changes
5. Provide custom arguments when running non-linear or bit-vector queries

Items 4 and 5 are tricky because cvc5 expects all options to be set either on the command line or at the start of the SMT file.  I don't think 4 will be easily addressable on the Verus side, so I'm hoping it might be feasible to change in cvc5.  Item 5 might be fixable on the Verus side, however, since we already create a new SMT context for non-linear and bit-vector queries.  I think we would need to modify `new_air_context_with_bucket_context` so we can pass in the reason for the spinoff here:
https://github.com/verus-lang/verus/blob/55b24eea93ad3dc907a4219e8b58a2f5f3f43afb/source/rust_verify/src/verifier.rs#L1389
and then add the requisite options at the start of the SMT interaction, rather than doing it at the last minute inside VIR's `sst_to_air`.

I'd be interested in thoughts on the items above, as well as the current refactoring.

### CODE REVIEW

**Comment by utaal** [comment]
(Marking as draft so I don't keep opening it when doing triage.)

File: source/vir/src/prelude.rs [hunk 1]
```diff
@@ -53,6 +55,21 @@ pub(crate) fn prelude_nodes(config: PreludeConfig) -> Vec<Node> {
...
+            :pattern (([height_lt] x y))
+            :qid prelude_height_lt
+            :skolemid skolem_prelude_height_lt
+            )))),
+        SmtSolver::Cvc5 => nodes_vec!(
+                    (declare-fun partial-order (Height Height) Bool)
+                    (axiom (forall ((x Height)) (partial-order x x)))
+                    (axiom (forall ((x Height) (y Height)) (=> (and (partial-order x y) (partial-order y x)) (= x y))))
+                    (axiom (forall ((x Height) (y Height) (z Height)) (=> (and (partial-order x y) (partial-order y z)) (partial-order x z))))
+                    (axiom (forall ((x Height) (y Height)) (! (= (height_lt x y) (and (partial-order x y) (not (= x y)))))))),
```
Comment (Chris-Hawblitzel) [inline source/vir/src/prelude.rs:71]: The "!" looks unused.

**Review by Chris-Hawblitzel — APPROVED** [review]

File: source/air/src/printer.rs [hunk 2]
```diff
@@ -351,7 +357,11 @@ impl Printer {
                 })
             }))
         }));
-        node!((declare-datatypes {decls} {defns}))
+        if datatypes.len() > 0 {
+            node!((declare-datatypes {decls} {defns}))
+        } else {
```
[the remark this answers is not in the document]
Comment (parno) [reply source/air/src/printer.rs]: Good point; I've updated it to be:
```
        if self.print_as_smt && datatypes.len() == 0 {
            // cvc5 doesn't like empty declare-datatypes
            nodes!(assert true)
        } else {
            node!((declare-datatypes {decls} {defns}))
        }
```

File: source/air/src/typecheck.rs [hunk 3]
```diff
@@ -130,14 +131,15 @@ fn get_bv_width(et: &Typ) -> Result<u32, TypeError> {
 }
 
 fn check_bv_unary_exprs(
+    solver: &SmtSolver,
```
[the remark this answers is not in the document]
Comment (parno) [reply source/air/src/typecheck.rs]: Good idea!  That's much cleaner.

File: source/vir/src/prelude.rs (hunk 1 above)
Comment (parno) [reply source/vir/src/prelude.rs]: Good point; I removed it.

**Comment by parno** [comment]
@Chris-Hawblitzel Thanks for reviewing the PR.  Do you think we need to address any or all of the issues I listed above before merging?

**Comment by Chris-Hawblitzel** [comment]
> @Chris-Hawblitzel Thanks for reviewing the PR. Do you think we need to address any or all of the issues I listed above before merging?

I think you can go ahead and squash and merge.  It's still largely an experimental feature, so it doesn't have to be feature-complete to be useful.

**Pull request merged**